### ISSUE TITLE
RTS DTR auto boot enter?

### ISSUE THREAD

**Issue by NoNamedCat** [wagiminator/MCU-Flash-Tools#2]
Hello again.
It is possible to implement this type of schema for the boot-loader in the code?

![image](https://github.com/wagiminator/MCU-Flash-Tools/assets/39467683/5cfa2780-3084-470a-b8c6-04606992cbd6)

I'm designing a new board for the PY32F002A micro-controller, wondering about using the CH340X as a USB-Serial communication IC for the board. If this could be implemented, I think is a no brainer to choose that IC instead of the CH340N (That doesn't have DTR or RTS)

**Comment by NoNamedCat** [comment]
This work well with the ESP8266 or the ESP32.
The EspTool (The program to burn the firmware for the ESP series) handle the reset line and the boot pin.
Do you think that is possible to implement this?

**Comment by wagiminator** [comment]
In PySerial you can directly set the control lines, e.g. "self.rts = True" (which means LOW on the RTS pin). So you can customize the software the way you want. By default RTS and DTR will be pulled low as soon as the communication begins. Keep in mind that this way of connection WILL cause problems, if you want to use it also for normal serial communication using a different serial monitor. By the way, the CH340N has an RTS pin. Take a look at the CH340E, which is quite small and has RTS, DTR and an LED pin.

**Comment by NoNamedCat** [comment]
Maybe adding a contraption based on transistor will help with this issue?
Like the DEV boards for the esp32 micro-controllers do?
Something like this:
![Screenshot_2023-11-09_15-33-41](https://github.com/wagiminator/MCU-Flash-Tools/assets/39467683/48913c6a-8619-4f41-8f25-a0bdb3c448d7)

**Comment by NoNamedCat** [comment]
The CH340X is designed for not needing that transistor contraption...
That's why i think about using that usb-serial IC.
Please see chapter 5.3. DTR and multi-mode MCU download

Of this datasheet: https://cdn.sparkfun.com/assets/5/0/a/8/5/CH340DS1.PDF

**Comment by wagiminator** [comment]
This looks like a clever solution. Just give it a try!

**Comment by wagiminator** [comment]
I added the additional control functions of the CH340X for BOOT0 and nRST according to the datasheet, but I can't test it at the moment. If you have implemented your board, please give me feedback. At the same time, I ordered a few CH340Xs, but I probably won't get around to experimenting with them until next year.

**Comment by NoNamedCat** [comment]
Hi! I really appreciate this. I'm working on the design first. I think I will finish it this weekend. I work in a company that buys PCBs from jlcpcb and normally I wait till the day that they buy many boards to not pay a cent for the shipping. So I don't know when I will get the boards in my hands. I really appreciate your work with all of this MCUs. If you want the design that I'm working on please just let me know. The board is double sided but the components are only on one side. You can select the working voltage with a jumper (3.3v or 5v) it has an onboard 3.3v regulator, USB type c with 5.1k resistors so it can work with USB type c to USB type c cables to trigger the host mode.

**Comment by wagiminator** [comment]
I am very curious already!

**Comment by NoNamedCat** [comment]
[PY32F002AF15P.zip](https://github.com/wagiminator/MCU-Flash-Tools/files/13325040/PY32F002AF15P.zip)
![Screenshot_2023-11-11_03-43-39](https://github.com/wagiminator/MCU-Flash-Tools/assets/39467683/1d4c532b-60d1-4e05-8ced-95fb7c567c66)
![Screenshot_2023-11-11_03-44-15](https://github.com/wagiminator/MCU-Flash-Tools/assets/39467683/0db1d545-dabc-4364-a3fd-a2adbaf8d10d)

**Comment by NoNamedCat** [comment]
In the zip file you can find the schematic, KiCad project (if you want to modify it), many .zip files containing all the gerber files for production in many PCB houses (jlcpcb, pcbway, fusionpcb, elecrow). It also have an interactive BOM...

I hope you like this.

**Comment by wagiminator** [comment]
Cool! I made some changes to my DevBoard as well and gave an order to JLCPCB. Let's see when everything arrives and I find the time to try it out.

**Comment by wagiminator** [comment]
Hi, CH340X and board are here. I was able to do a few tests. ISP software works - so far so good. Apart from that, unfortunately the problems I feared arise. If you try to use the CH340X as a "normal" USB-to-serial converter with any serial monitor, the PY32 will be reset because, as usual, the RTS pin is automatically set to LOW when data is sent. Second problem is that despite the pulldown, BOOT0 is pulled to HIGH by default by the CH340X. A normal reset using a button is therefore not possible - you always end up in the bootloader. I currently don't know how to work around the problem...

**Comment by NoNamedCat** [comment]
Hello Wagiminator. Maybe we can add some components to make the thing works. Can you share the schematic of your design so I can analyze it?

**Comment by NoNamedCat** [comment]
maybe with some capacitors and transistors we can make it works

**Comment by NoNamedCat** [comment]
Also... Do you have a logic analyzer or some scope that can help us to debug the DTR and RST lines?

**Comment by wagiminator** [comment]
Yes, I have a logic analyzer and I also used it here. You can certainly create a solution with additional components, but of course they make the whole thing more complex and then they also have to fit on the board.

**Comment by wagiminator** [comment]
![ch340x](https://github.com/wagiminator/MCU-Flash-Tools/assets/62595727/43b0ec8e-218e-49ce-9f8f-64242805f65a)

**Comment by NoNamedCat** [comment]
A resistor and capacitor on the BOOT0 will add a delay in the reset line of the CH340X. It may be the solution on the  boot state of your board?

**Comment by NoNamedCat** [comment]
https://www.build-electronic-circuits.com/rc-delay-element/

**Comment by NoNamedCat** [comment]
so... If you add that. you need to make a delay in time on the code for programming the device. You need to put the DTR line in the level for flashing state. Then you add a delay so the capacitor will reach the logic level of the boot state. Then you reset the MCU. The solution is two small components. What do you think?

**Comment by wagiminator** [comment]
That won't help. The situation looks like this: If the board is connected via USB, then RTS and DTR are high. If I open the connection to the CH340X for example with CuteCom, RTS and DTR are immediately set to low and remain low until I close the connection.

**Comment by wagiminator** [comment]
I suspect that the CH340X is pulling the DTR line hard to VCC (open-source), if I overpress that there will probably be a short.

**Comment by NoNamedCat** [comment]
Can you try it with the serial debugger in Arduino IDE?

**Comment by NoNamedCat** [comment]
Also... Here is a link explaining the CH340X IC:
https://www.youtube.com/watch?v=CGLHfyW0rYE

**Comment by NoNamedCat** [comment]
Have you tried to connect T-NOW to CTS with a 4.7K resistor instead of using the 4.7 pulldown resistor on the T-NOW?

**Comment by wagiminator** [comment]
It's the same with the Arduino IDE (at least on Linux, if that has any influence). Thanks for the video link. I think it works great if you make a programming device with the chip. However, the problems remain for a full on-board solution. Connecting the resistor between DTR and CTS makes no sense here. DTR is then only set push-pull instead of open-source. This certainly makes sense with other MCUs.

**Comment by NoNamedCat** [comment]
I feel sad about that. Sorry to hear that the solution didn't work.

**Comment by wagiminator** [comment]
Yes, that's a real shame. It would have been a really elegant solution. Well, maybe I can think of something else.

**Comment by wagiminator** [comment]
If DTR were LOW in its default state, I would immediately have a simple solution. It's really not clear to me why this is solved this way with the chip.

**Comment by NoNamedCat** [comment]
Maybe if you put a capacitor like a DC filter on the rest line? I don't want to lose hope :c

**Comment by wagiminator** [comment]
This would be the easy solution ... if DTR were LOW on default.

**Comment by NoNamedCat** [comment]
Hello again Wagiminator. Hope you are fine.
I want to tell you about the new repo of [HalfSweet](https://github.com/HalfSweet) (the creator of the PY32DUINO).
She designed a board for the PY32F030 MCU and i think she have done a solution based on transistors like this:
![Screenshot_2023-11-28_17-10-39](https://github.com/wagiminator/MCU-Flash-Tools/assets/39467683/30fda648-386c-4986-b20f-c4a433f29578)

**Comment by NoNamedCat** [comment]
Here is the repo:
https://github.com/HalfSweet/PY32F030K28U_Arduino_Nano

**Comment by NoNamedCat** [comment]
I think that you can replace the mosfets for NPN transistors like the 2n2222 or 2n3904

**Comment by NoNamedCat** [comment]
it's the same circuit that i posted before_
![image](https://github.com/wagiminator/MCU-Flash-Tools/assets/39467683/2f87f9b5-b4e9-4cd5-8cb1-354f6e6304ee)
but it uses the third transistor to negate the BOOT0 state

**Comment by NoNamedCat** [comment]
The ch340x function and behavior is like having the two first transistors, do you think that if we put the third transistor only on the boot line to negate the state, we can manage to make the board work?

**Comment by wagiminator** [comment]
When I look at the behavior of the CH340X, I come to the conclusion that the only difference to the CH340E is that the DTR signal can be ouput on a pin if necessary. This means that the circuit with the transistors is necessary here if you want to use the chip also as a USB-to-serial converter and not just as a programming device.
In principle, you can also build the circuit with BJTs instead of MOSFETs. You then need resistors for the base. The body diode of the MOSFETs might also have to be taken into account.

**Comment by NoNamedCat** [comment]
If I make the board design with all of this stuff, I will post it on this topic. Are you still interested? It will be my pleasure to give you that

**Comment by wagiminator** [comment]
Oh yes, I'm looking forward to it.

**Comment by NoNamedCat** [comment]
I finish the board design. I grabbed the transistors for the boot pins from the repo above. Can you tell me what do you think about it? There is a space left on the board that I can use for some leds on the TX and RX pins (Like the Arduino boards).
All the passive elements (Caps and Resistors) are 0805, smaller ones are difficult for me.
If you want some personalized silkscreen I can do it for you (And this repo)

![PY32F002](https://github.com/wagiminator/MCU-Flash-Tools/assets/39467683/fa73e616-0223-4512-9704-63ab4107ba3b)

**Comment by NoNamedCat** [comment]
![Screenshot_2023-11-29_17-54-29](https://github.com/wagiminator/MCU-Flash-Tools/assets/39467683/15659c60-4cc6-4721-818e-bad7067f668b)
![Screenshot_2023-11-29_17-55-58](https://github.com/wagiminator/MCU-Flash-Tools/assets/39467683/98cae8ab-a80c-4f99-8c55-1d0b5e8ad0c7)

**Comment by NoNamedCat** [comment]
![bottom](https://github.com/wagiminator/MCU-Flash-Tools/assets/39467683/ef6691f3-2f18-42b9-80a8-6207aa35cb64)
![top](https://github.com/wagiminator/MCU-Flash-Tools/assets/39467683/71b7f527-1135-48a1-b1d9-cf61c1132238)

**Comment by wagiminator** [comment]
Nice Logo! Is there a reason why you used CTS instead of DTR? I don't think that will work.

**Comment by wagiminator** [comment]
There might be som problems with the CH340X if you select 3V3 as MCU VCC. If you power the CH340X with 3V3 on the VCC pin, you should also power it with 3V3 on the 3V pin, I think. You can solve that with a permanent connection of the 3V pin to +3.3V of the LDO.

**Comment by wagiminator** [comment]
Please take a closer look at the MOSFET circuit and compare it with that of HalfSweet. How do you want to pull the NRST pin to LOW?

**Comment by NoNamedCat** [comment]
Tks for the hints. I will correct those and share the design with you.

**Comment by NoNamedCat** [comment]
I will need to redo many things with the changes that you proposed. But is very late here right now. Tomorrow I will continue to design the board. If you like my logo... I can design one for you if you want.

**Comment by NoNamedCat** [comment]
![Screenshot_2023-11-30_02-04-31](https://github.com/wagiminator/MCU-Flash-Tools/assets/39467683/8d84a251-cb07-4cc2-b082-3cd49fa6564f)

**Comment by NoNamedCat** [comment]
Here is the project file, including design files, PDF for schematic, interactive BOM in html, CSV BOM file, and gerber files for production in many sites.

[PY-CAT-002.zip](https://github.com/wagiminator/MCU-Flash-Tools/files/13515961/PY-CAT-002.zip)

**Comment by wagiminator** [comment]
I still believe you have to swap PF2 and PA6 as well as DTR and RTS at the MOSFET circuit.

**Comment by wagiminator** [comment]
If you have some room left on the board, some 100nF capacitors for the power pins of the CH340 and PY32 will make the whole thing a little more stable, although it certainly works without it.

**Comment by NoNamedCat** [comment]
Ok. I swapped the DTR and RTS lines. But I don't understand why we need to swap the RX and TX pins on the board.
With serial ports you need to connect the RX to TX pin and the TX to the RX pin (The TX pin on the CH320 transmit, and the PY32F002 PA3 receive).

**Comment by NoNamedCat** [comment]
The RX and TX lines works fine for me.

**Comment by NoNamedCat** [comment]
I have a problem with the current design. I want to make a panalization (Array of boards) of the design. The USB port and the buttons for reset and boot crash in the panel. The idea is to make a panel of 3*2 on the standard size of the JLCPCB. If the board works i think that we don't need the buttons in the board. Do you think that we need the buttons?

**Comment by wagiminator** [comment]
I never mentioned TX/RX. You should swap PF2 (NRST) PB6 (BOOT0) on the MOSFET circuit (see HalfSweet's circuit).

Well, if it works you don't necessarily need the buttons. If necessary, you can connect some via the pinouts.

I'll probably have to adjust the ISP software...

**Comment by NoNamedCat** [comment]
I'm sorry for the spam. This design would not exist if it weren't for you. Tks a lot.

**Comment by HalfSweet** [comment]
Actually it's a very simple question...the condition for py32 to enter the bootloader is very simple, just have the boot0 pin high at reset. So we may easily find a similar schematic online.
![图片](https://github.com/wagiminator/MCU-Flash-Tools/assets/60973476/5ec71521-4bfc-49a6-9407-007048ee7ae1)
But there is a problem, when we switch the serial port on and off, the level of DTR and RTS may not be fixed, which may cause us to enter a different mode than expected by mistake.
And in the Arduino IDE, DTR and RTS are set to true when the serial port monitor is turned on - a very bad design, and to avoid this problem, back when I was designing the [Air001 development board](https://air001.cn) I included the An ESP32 autodownload-like design using two BJT/MOS to implement a hardware heterodyne circuit, which we can see in <https://gitee.com/openLuat/luatos-broads/tree/master/broads/Air001-Board> at the time.

Of course there is a slight problem with this circuit, which means we can't control the levels of boot0 and RST at the same time, so we compensated for this by connecting a capacitor in parallel to GND for both boot0 and RST, and performing pull-downs and pull-ups according to the needs of the running app.

At [AirISP](https://github.com/Air-duino/AirISP), I implemented the automatic download logic for the relevant circuits

**Comment by NoNamedCat** [comment]
You are a FullSweet ;) tks a lot for sharing your knowledge with me. Can I share this conversation with Wagiminator so he can design his board too?

**Comment by HalfSweet** [comment]
> You are a FullSweet ;) tks a lot for sharing your knowledge with me. Can I share this conversation with Wagiminator so he can design his board too?

Of course, our sessions are open and we can have discussions together if we can
